From 9c503721c49feb3a5d774c3d8d789832cbd973fa Mon Sep 17 00:00:00 2001
From: Wenbin Song <wenbin.song@nxp.com>
Date: Wed, 30 Jun 2021 16:38:08 +0800
Subject: [PATCH 16/18] enetc: adjust the link status dynamically for ec_master

Because the ec_enetc_vf module works in phy-less mode,
ec_master can not be notified dynamically when the link status changed.

Signed-off-by: Wenbin Song <wenbin.song@nxp.com>
---
 devices/enetc/enetc.c    | 1 -
 devices/enetc/enetc_pf.c | 5 +++++
 devices/enetc/enetc_vf.c | 4 ++++
 3 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/devices/enetc/enetc.c b/devices/enetc/enetc.c
index 8eb19b3..3e04a67 100644
--- a/devices/enetc/enetc.c
+++ b/devices/enetc/enetc.c
@@ -1095,7 +1095,6 @@ int enetc_open(struct net_device *ndev)
 
 	enetc_setup_bdrs(priv);
 	enetc_start(ndev);
-	ecdev_set_link(priv->ecdev, 1);
 	return 0;
 
 err_alloc_rx:
diff --git a/devices/enetc/enetc_pf.c b/devices/enetc/enetc_pf.c
index 778db4b..3a048d4 100644
--- a/devices/enetc/enetc_pf.c
+++ b/devices/enetc/enetc_pf.c
@@ -477,6 +477,7 @@ static void enetc_pl_mac_link_up(struct phylink_config *config,
 		enetc_sched_speed_set(priv, speed);
 
 	enetc_mac_enable(&pf->si->hw, true);
+	ecdev_set_link(priv->ecdev, 1);
 }
 
 static void enetc_pl_mac_link_down(struct phylink_config *config,
@@ -484,8 +485,12 @@ static void enetc_pl_mac_link_down(struct phylink_config *config,
 				   phy_interface_t interface)
 {
 	struct enetc_pf *pf = phylink_to_enetc_pf(config);
+	struct enetc_ndev_priv *priv;
+
+	priv = netdev_priv(pf->si->ndev);
 
 	enetc_mac_enable(&pf->si->hw, false);
+	ecdev_set_link(priv->ecdev, 0);
 }
 
 static const struct phylink_mac_ops enetc_mac_phylink_ops = {
diff --git a/devices/enetc/enetc_vf.c b/devices/enetc/enetc_vf.c
index 43bbaf4..458b129 100644
--- a/devices/enetc/enetc_vf.c
+++ b/devices/enetc/enetc_vf.c
@@ -123,6 +123,10 @@ static int enetc_vf_probe(struct pci_dev *pdev,
 		goto err_reg_ec_net;
 	}
 
+	/* The vf driver works on phy-less mode, so it can not be
+	 * notified when the link status changed. */
+	ecdev_set_link(priv->ecdev, 1);
+
 	netif_info(priv, probe, ndev, "%s v%s\n",
 		   enetc_drv_name, enetc_drv_ver);
 
-- 
2.25.1

